### ISSUE TITLE
1.3 - > 2.0 UserManager's Entity Manager not synced with switched connection to different database

### ISSUE THREAD

**Issue by Mopster** [FriendsOfSymfony/FOSUserBundle#2852]
I'm working on upgrading from version 1.3 to 2.x and encountered the following behavior : 

For some explanation of my setup : 

I have a multi-tenancy app which switches database connection based on the subdomain. This is done through a listener on kernel request. The default entity manager gets updated to connect to correct database. 

The issue I'm having now is when loading fixtures, we have an entity manager called 'fixtures' that connects to the fixtures database we use for development and testing purposes.

`php app/console doctrine:fixtures:load --em=fixtures`

The problem I'm experiencing now is the fixtures are being loaded in the default database and not in the 'fixtures' database. When debugging, I noticed the connection of the Object Manager in the User Manager is still the old connection and connects to the wrong database. While the Entity Manager in the container has the correct connection. 

In our 1.3 implementation of the FOSUserBundle we extended the UserManager and set the ObjectManager before running the fixtures and also in the Listener mentioned above. Checking the history of this code, this was done intentionally since "the User Manager maintains its own instance of the entity manager". (this part is +5y old). 

```
    public function setObjectManager(ObjectManager $om)
    {
        $this->objectManager = $om;
        $this->repository = $om->getRepository($this->class);

        $metadata = $om->getClassMetadata($this->class);
        $this->class = $metadata->getName();
    }
```

In 2.0, this no longer works. And i'm now looking for a solution. I've been breaking my head on it for several hours this week and looking for a way out. Getting rid of the 'hack' would be nice or find a cleaner workaround. Or help me understand how to make sure the UserManager uses the updated Entity Manager and connection.

I tried working directly with version 2.1.2, but since I noticed this issue, I went back to 2.0.0-alpha3 and the problem remained unless I revert back to 1.3

I found an old thread of my former colleague here : https://github.com/FriendsOfSymfony/FOSUserBundle/issues/413

**Comment by stof** [comment]
`--em=fixtures` in `doctrine:fixtures:load` changes the entity manager used by the fixtures loader. It does not change the default entity manager, and it does not change the service definitions in your container. Any service were the entity manager gets injected would still use the one it is using.

And mutating a service after it has been instantiated to change its dependencies is something which is not guaranteed to work.

If you want to use a separate database and have all your services use it, a better solution could be to use a separate environment and change the database used by the default entity manager, rather than having 2 managers and trying to change dynamically which one is used by all services.

**Comment by Mopster** [comment]
ok this took me some time to figure out, but I got it, thanks for the explanation above!

I moved my `fixtures` connection and entity manager config to the `test` environment, set them both to default to `fixtures` and told the fixtures loader to load the `test` environment.

The fixtures are being loaded in the correct database and my `dev` environment can connect it using the default entity manager which is reloaded using an onKernelRequest listener.

It's still not the suggested approach. Which would mean a env file per database ? Isn't there a clean way of setting the correct entity manager/connection before it gets created, based on the subdomain? (But this question is off topic for this bundle, so I'll close this issue as it has been resolved)
